### ISSUE TITLE
How to close bidirectional stream gracefully without having `*stats.End` filled with `context.Canceled` error

### ISSUE THREAD

**Issue by vitalyisaev2** [grpc/grpc-go#5164]
Hello, please clarify the process of a bidirectional stream graceful termination from the side of the client.

In [docs](https://pkg.go.dev/google.golang.org/grpc#ClientConn.NewStream), it is stated that there are 4 options for a stream termination:

> 1. Call Close on the ClientConn.

This is not an option since the connection is generally reused among streams.

> 2. Cancel the context provided.

This method is pretty straightforward, but the problem is that in this case `*stats.End` instances passed to `stats.Handler.HandleRPC()` on the client side will always have `Error` field filled with `context.Cancelled` error. This makes it impossible to distinguish true context cancellations from the situation when the context was just used for a graceful stream termination. By "true context cancellation" I mean cases, when the upstream client has left (for example, user closed his browser, and the context was canceled recursively everywhere down to backend service).

As long as `stats.Handler` was designed for client- and server-side statistics accumulation, the problem described above leads to the increase of false errors in client-side stats. Every terminated stream will increment the error counter.

> 3. Call RecvMsg until a non-nil error is returned. A protobuf-generated
>   client-streaming RPC, for instance, might use the helper function
>   CloseAndRecv (note that CloseSend does not Recv, therefore is not
>   guaranteed to release all resources).

For a bidirectional stream, `CloseAndRecv` is not available, whereas `CloseSend` does not provide guarantees.

> 4. Receive a non-nil, non-io.EOF error from Header or SendMsg.

Seems like this may only happen if server or network have failed, so it cannot be applied in general case.

In summary, the only reliable way to close bidirectional stream is to cancel stream context explicitly (`runRouteChat` from examples [confirms that](https://github.com/grpc/grpc-go/blob/master/examples/route_guide/client/client.go#L119)). But this approach causes the registration of false errors within `stats.Handler`. 

Are there any mistakes in my reasoning? If no, are there any workarounds for this issue? Thank you.

**Comment by menghanl** [comment]
> For a bidirectional stream, `CloseAndRecv` is not available, whereas `CloseSend` does not provide guarantees.

1. If the client calls `CloseSend`, the server's `Recv()` will get `io.EOF`.
2. The server `return nil` when it receives `io.EOF`
3. The client's `Recv()` will eventually receive `io.EOF` (Note that messages are delivered in order, so the client should keep calling Recv until until a non-nil error is returned)

`RouteChat` in the route guide example shows this:
https://github.com/grpc/grpc-go/tree/master/examples/route_guide

**Comment by vitalyisaev2** [comment]
@menghanl Thanks a lot, now it's clear
